pkg/waifu_vault_api.go

```diff
@@ -70,19 +75,27 @@ func (re *Api) UploadFile(options mod.WaifuvaultPutOpts) (*mod.WaifuResponse[str
 		"password":      options.Password,
 	}, "")
 
-	r, err := createRequest(http.MethodPut, uploadUrl, &body, writer)
+	r, err := re.createRequest(http.MethodPut, uploadUrl, &body, writer, ctx)
 	if err != nil {
 		return nil, err
 	}
-	resp, err := client.Do(r)
+	resp, err := re.client.Do(r)
 	if err != nil {
 		return nil, err
 	}
 	return getResponse[string](resp)
 }
 
-func createRequest(method, url string, body io.Reader, writer *multipart.Writer) (*http.Request, error) {
-	r, err := http.NewRequest(method, url, body)
+func (re *api) createRequest(method, url string, body io.Reader, writer *multipart.Writer, ctx *context.Context) (*http.Request, error) {
+	var r *http.Request
+	var err error
+
+	if ctx != nil {
+		r, err = http.NewRequestWithContext(*ctx, method, url, body)
+	} else {
+		r, err = http.NewRequest(method, url, body)
+	}
+
 	if err != nil {
 		return nil, err
 	}
```

Comment:
With my previous comments, make `ctx` not pointer, then you can simplify this code.

```
r, err  := http.NewRequestWithContext(ctx, method, url, body)

if err != nil {
		return nil, err
}
```